client/build.go

```diff
@@ -0,0 +1,97 @@
+package client
+
+import (
+	"context"
+
+	"github.com/moby/buildkit/client/buildid"
+	gateway "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/moby/buildkit/frontend/gateway/grpcclient"
+	gatewayapi "github.com/moby/buildkit/frontend/gateway/pb"
+	"github.com/moby/buildkit/session"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+type build struct {
+	id string
+	s  *session.Session
+	c  gateway.Client
+}
+
+func (c *Client) Build(ctx context.Context, opt SolveOpt, build gateway.BuildFunc, statusChan chan *SolveStatus) (*SolveResponse, error) {
+	defer func() {
+		if statusChan != nil {
+			close(statusChan)
+		}
+	}()
+
+	if opt.Frontend != "" {
+		return nil, errors.New("invalid build opts cannot use frontend")
+	}
+
+	feOpts := opt.FrontendAttrs
+	opt.FrontendAttrs = nil
+
+	workers, err := c.ListWorkers(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "listing workers")
+	}
+	var gworkers []gateway.WorkerInfo
+	for _, w := range workers {
+		gworkers = append(gworkers, gateway.WorkerInfo{
+			ID:        w.ID,
+			Labels:    w.Labels,
+			Platforms: w.Platforms,
+		})
+	}
+
+	cb := func(ref string, s *session.Session) error {
+		g, err := grpcclient.New(ctx, feOpts, s.ID(), "", c.gatewayClient(ref), gworkers)
```

Comment:
I pass product as `""` on the logic that the `build` callback is in the process so it surely already knows whatever it needs to know, but maybe this should be an argument to this `Build` method?